### PR TITLE
JAMES-3955 Better handle and avoid RabbitMQ message processing timeouts

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java [hunk 1]
```diff
@@ -133,7 +137,7 @@ private Mono<RabbitMQMailQueueItem> loadItem(AcknowledgableDelivery response) {
                 LOGGER.error("Failed to load email, requeue corresponding message", e);
                 response.nack(REQUEUE);
                 return Mono.empty();
-            });
+            }).timeout(TIMEOUT);
```
Comment (Arsnael) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java:140]: Does that really work though? Don't we nack the message already anyways to rabbitmq before the timeout occurs? Trying to understand :)

Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java:140]: Timeout generates an error that generates the nack

I'll take advantage of boring meetings to test if you want...

Comment (Arsnael) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java:140]: Would love to be able to test that yes :)

Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java:140]: Manually tested:

```
    @Test
    void registerGroupShouldCreateRetryExchange() throws Exception {
        GroupA registeredGroup = new GroupA();
        CountDownLatch countDownLatch = new CountDownLatch(1);
        EventListener listener = new EventListener.GroupEventListener() {
            private final AtomicInteger atomicInteger = new AtomicInteger(0);

            @Override
            public Group getDefaultGroup() {
                return registeredGroup;
            }

            @Override
            public void event(Event event) throws Exception {
                System.out.println("Processing...");
                if (atomicInteger.getAndIncrement() == 0) {
                    System.out.println("Waiting...");
                    Thread.sleep(Duration.ofMinutes(2).toMillis());
                }
                System.out.println("success");
                countDownLatch.countDown();
            }
        };
        eventBus.register(listener, registeredGroup);

        eventBus.dispatch(EVENT, KEY_1).block();
        eventBus.dispatch(EVENT, KEY_1).block();
        System.out.println("events submitted");
        countDownLatch.await();
    }
```

And I did set the timeout on James side to 1 minute.

```
events submitted
Processing...
Waiting...
07:07:51.520 [ERROR] o.a.j.e.GroupConsumerRetry - Exception happens when handling event after 0 retries
java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 60000ms in 'contextWrite' (and no fallback has been configured)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:295)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280)
	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419)
	at reactor.core.publisher.FluxOnErrorReturn$ReturnSubscriber.onNext(FluxOnErrorReturn.java:162)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
07:07:51.523 [ERROR] r.c.p.Operators - Operator called default onErrorDropped
java.lang.InterruptedException: sleep interrupted
	at java.base/java.lang.Thread.sleep(Native Method)
	at org.apache.james.events.RabbitMQEventBusTest$1.event(RabbitMQEventBusTest.java:310)
	at org.apache.james.events.EventListener$ReactiveWrapper.lambda$reactiveEvent$0(EventListener.java:66)
	at com.github.fge.lambdas.runnable.RunnableChainer.doRun(RunnableChainer.java:18)
	at com.github.fge.lambdas.runnable.ThrowingRunnable.run(ThrowingRunnable.java:16)
	... 10 common frames omitted
Wrapped by: com.github.fge.lambdas.ThrownByLambdaException: java.lang.InterruptedException: sleep interrupted
	at com.github.fge.lambdas.runnable.ThrowingRunnable.run(ThrowingRunnable.java:20)
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:73)
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:32)
	at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:227)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Processing...
success
```

Which does the job ;-)

**Review by Arsnael — APPROVED** [review]

**Pull request merged**